exonum-java-binding/core/rust/src/testkit/mod.rs

```diff
@@ -213,3 +214,95 @@ fn create_java_keypair<'a>(
         &[secret_key_byte_array.into(), public_key_byte_array.into()],
     )
 }
+
+// Converts Java array of `TestKitServiceInstances` to vector of `InstanceConfig`.
+fn instance_configs_from_java_array(
+    env: &JNIEnv,
+    services: jobjectArray,
+) -> JniResult<Vec<InstanceConfig>> {
+    let mut instance_configs = vec![];
+    let num_services = env.get_array_length(services)?;
+    for i in 0..num_services {
+        let service_obj = env.get_object_array_element(services, i)?;
+
+        let artifact_id = convert_to_string(
+            &env,
+            env.get_field(service_obj, "artifactId", "Ljava/lang/String;")?
+                .l()?,
+        )?;
+        let artifact_filename = convert_to_string(
+            &env,
+            env.get_field(service_obj, "artifactFilename", "Ljava/lang/String;")?
+                .l()?,
+        )?;
+        let service_specs_obj: jobjectArray = env
+            .get_field(
+                service_obj,
+                "serviceSpecs",
+                "[Lcom/exonum/binding/testkit/ServiceSpec",
+            )?
+            .l()?
+            .into_inner();
+
+        let num_specs = env.get_array_length(service_specs_obj)?;
+
+        for k in 0..num_specs {
+            let service_spec = env.get_object_array_element(service_specs_obj, k)?;
+
+            let service_id = env.get_field(service_spec, "serviceId", "I")?.i()? as u32;
+            let service_name = convert_to_string(
```

Comment:
Code is duplicated between this function and `time_service_specs_from_java_array`

Reply:
Yes, I also noticed that but didn't end up with a quick and easy implementation for the moment.

Reply:
Refactored.